File: providers/openweathermap.py
"""
OpenWeatherMap provider for global weather data.


Provides weather data for any location worldwide via OpenWeatherMap API.
Free tier: 1000 calls/day.
"""

import urllib.request
import urllib.parse
import json
from datetime import datetime, timezone
from typing import Optional

from ..models import WeatherData, WeatherCondition, Location
from .base import WeatherProvider, ProviderError


class OpenWeatherMapProvider(WeatherProvider):
    """
    OpenWeatherMap weather provider.

    Coverage: Global
    API Key: Required (free tier: 1000 calls/day)
    Priority: 10 (fallback for non-HK locations)
    """

    BASE_URL = "https://api.openweathermap.org/data/2.5"

    # OpenWeatherMap condition codes: https://openweathermap.org/weather-conditions
    CONDITION_MAP = {
        # Thunderstorm (2xx)
        200: WeatherCondition.THUNDERSTORM,
        201: WeatherCondition.THUNDERSTORM,
        202: WeatherCondition.THUNDERSTORM,
        210: WeatherCondition.THUNDERSTORM,
        211: WeatherCondition.THUNDERSTORM,
        212: WeatherCondition.THUNDERSTORM,
        221: WeatherCondition.THUNDERSTORM,
        230: WeatherCondition.THUNDERSTORM,
        231: WeatherCondition.THUNDERSTORM,
        232: WeatherCondition.THUNDERSTORM,
        # Drizzle (3xx)
        300: WeatherCondition.DRIZZLE,
        301: WeatherCondition.DRIZZLE,
        302: WeatherCondition.DRIZZLE,
        310: WeatherCondition.DRIZZLE,
        311: WeatherCondition.DRIZZLE,
        312: WeatherCondition.DRIZZLE,
        313: WeatherCondition.DRIZZLE,
        314: WeatherCondition.DRIZZLE,
        321: WeatherCondition.DRIZZLE,
        # Rain (5xx)
        500: WeatherCondition.DRIZZLE,  # Light rain
        501: WeatherCondition.RAIN,
        502: WeatherCondition.HEAVY_RAIN,
        503: WeatherCondition.HEAVY_RAIN,
        504: WeatherCondition.HEAVY_RAIN,
        511: WeatherCondition.RAIN,
        520: WeatherCondition.SHOWERS,
        521: WeatherCondition.SHOWERS,
        522: WeatherCondition.HEAVY_RAIN,
        531: WeatherCondition.SHOWERS,
        # Snow (6xx)
        600: WeatherCondition.SNOW,  # Light snow
        601: WeatherCondition.SNOW,
        602: WeatherCondition.HEAVY_SNOW,
        611: WeatherCondition.SLEET,
        612: WeatherCondition.SLEET,
        613: WeatherCondition.SLEET,
        615: WeatherCondition.SLEET,
        616: WeatherCondition.SLEET,
        620: WeatherCondition.SNOW,
        621: WeatherCondition.SNOW,
        622: WeatherCondition.HEAVY_SNOW,
        # Atmosphere (7xx)
        701: WeatherCondition.MIST,
        711: WeatherCondition.MIST,  # Smoke -> mist
        721: WeatherCondition.MIST,  # Haze -> mist
        731: WeatherCondition.MIST,  # Dust whirls -> mist
        741: WeatherCondition.FOG,
        751: WeatherCondition.MIST,  # Sand -> mist
        761: WeatherCondition.MIST,  # Dust -> mist
        762: WeatherCondition.MIST,  # Volcanic ash -> mist
        771: WeatherCondition.WINDY,
        781: WeatherCondition.THUNDERSTORM,  # Tornado
        # Clear (800)
        800: WeatherCondition.SUNNY,
        # Clouds (80x)
        801: WeatherCondition.PARTLY_CLOUDY,
        802: WeatherCondition.PARTLY_CLOUDY,
        803: WeatherCondition.CLOUDY,
        804: WeatherCondition.CLOUDY,
    }

    def __init__(self, api_key: str):
        """
        Initialize OpenWeatherMap provider.

        Args:
            api_key: OpenWeatherMap API key (get free at openweathermap.org)
        """
        self.api_key = api_key
        self._priority = 10  # Fallback priority

    @property
    def name(self) -> str:
        return "openweathermap"

    @property
    def priority(self) -> int:
        return self._priority

    @property
    def supports_forecast(self) -> bool:
        return True

    def supports_location(self, location: Location) -> bool:
        """OpenWeatherMap supports all global locations."""
        return True

    async def get_current(self, location: Location) -> WeatherData:
        """Fetch current weather from OpenWeatherMap."""
        params = {
            "q": location.normalized,
            "appid": self.api_key,
            "units": "metric",
        }
        url = f"{self.BASE_URL}/weather?{urllib.parse.urlencode(params)}"

        try:
            with urllib.request.urlopen(url, timeout=10) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise ProviderError("OpenWeatherMap API key is invalid")
            elif e.code == 404:
                raise ProviderError(f"Location not found: {location.raw}")
            else:
                raise ProviderError(f"OpenWeatherMap API error: HTTP {e.code}")
        except Exception as e:
            raise ProviderError(f"OpenWeatherMap request failed: {e}")

        # Also fetch today's forecast for high/low temps
        today_high = None
        today_low = None
        try:
            forecast_params = {
                "q": location.normalized,
                "appid": self.api_key,
                "units": "metric",
                "cnt": 8,  # Next 24 hours (3-hour intervals)
            }
            forecast_url = f"{self.BASE_URL}/forecast?{urllib.parse.urlencode(forecast_params)}"
            with urllib.request.urlopen(forecast_url, timeout=10) as resp:
                forecast_data = json.loads(resp.read().decode("utf-8"))

            # Get today's temps from forecast
            temps = []
            today = datetime.now(timezone.utc).date()
            for item in forecast_data.get("list", []):
                dt = datetime.fromtimestamp(item.get("dt", 0), tz=timezone.utc)
                if dt.date() == today:
                    temps.append(item.get("main", {}).get("temp"))

            if temps:
                today_high = max(temps)
                today_low = min(temps)
        except Exception:
            pass  # High/low are optional enhancements

        return self._parse_current(data, today_high, today_low)

    async def get_current_with_air_quality(self, location: Location) -> WeatherData:
        """Fetch current weather with air quality data."""
        # First get weather data
        weather_data = await self.get_current(location)

        # Get coordinates from weather response for air quality
        params = {
            "q": location.normalized,
            "appid": self.api_key,
            "units": "metric",
        }
        url = f"{self.BASE_URL}/weather?{urllib.parse.urlencode(params)}"

        try:
            with urllib.request.urlopen(url, timeout=10) as resp:
                data = json.loads(resp.read().decode("utf-8"))
            coords = data.get("coord", {})
            lat = coords.get("lat")
            lon = coords.get("lon")

            if lat and lon:
                air_quality = await self.get_air_quality(lat, lon)
                # Add air quality to weather data
                weather_data.aqi = air_quality.get("aqi")
                weather_data.pm25 = air_quality.get("pm25")
                weather_data.pm10 = air_quality.get("pm10")
                weather_data.o3 = air_quality.get("o3")
                weather_data.no2 = air_quality.get("no2")
        except Exception:
            pass  # Air quality is optional

        return weather_data

    async def get_forecast(
        self, location: Location, days: int = 3
    ) -> list[WeatherData]:
        """Fetch weather forecast from OpenWeatherMap."""
        # OpenWeatherMap 5-day forecast gives 3-hour intervals (40 readings)
        # We need to aggregate into daily forecasts
        params = {
            "q": location.normalized,
            "appid": self.api_key,
            "units": "metric",
            "cnt": min(days * 8, 40),  # 8 readings per day, max 40 (5 days)
        }
        url = f"{self.BASE_URL}/forecast?{urllib.parse.urlencode(params)}"

        try:
            with urllib.request.urlopen(url, timeout=10) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise ProviderError("OpenWeatherMap API key is invalid")
            elif e.code == 404:
                raise ProviderError(f"Location not found: {location.raw}")
            else:
                raise ProviderError(f"OpenWeatherMap API error: HTTP {e.code}")
        except Exception as e:
            raise ProviderError(f"OpenWeatherMap request failed: {e}")

        return self._parse_forecast(data, days)

    def _parse_current(self, data: dict, today_high: float | None = None, today_low: float | None = None) -> WeatherData:
        """Parse OpenWeatherMap current weather response."""
        weather = data.get("weather", [{}])[0]
        main = data.get("main", {})
        wind = data.get("wind", {})
        clouds = data.get("clouds", {})
        rain = data.get("rain", {})
        snow = data.get("snow", {})

        condition_code = weather.get("id", 800)
        condition = self.CONDITION_MAP.get(condition_code, WeatherCondition.UNKNOWN)

        # Wind direction
        wind_deg = wind.get("deg")
        wind_direction = self._deg_to_direction(wind_deg) if wind_deg else None

        # Format wind description (e.g., "12 km/h NE")
        wind_description = None
        if wind.get("speed"):
            wind_speed_kmh = wind["speed"] * 3.6  # Convert m/s to km/h
            if wind_direction:
                wind_description = f"{wind_speed_kmh:.0f} km/h {wind_direction}"
            else:
                wind_description = f"{wind_speed_kmh:.0f} km/h"

        # Calculate rain probability from cloud cover (rough estimate)
        rain_probability = None
        if clouds.get("all") is not None:
            rain_probability = min(100, clouds["all"])

        # Capitalize description
        description = weather.get("description", "")
        if description:
            description = description.capitalize()

        return WeatherData(
            location=data.get("name", "Unknown"),
            temperature=main.get("temp"),
            feels_like=main.get("feels_like"),
            temp_high=today_high,
            temp_low=today_low,
            humidity=main.get("humidity"),
            pressure=main.get("pressure"),
            wind_speed=wind.get("speed"),
            wind_direction=wind_direction,
            wind_description=wind_description,
            condition=condition,
            description=description,
            precipitation_chance=rain_probability,
            observed_at=datetime.now(timezone.utc),
            provider_name=self.name,
        )

    def _parse_forecast(self, data: dict, days: int) -> list[WeatherData]:
        """Parse OpenWeatherMap 5-day forecast into daily forecasts."""
        # Group readings by date
        daily_data = {}
        location_name = data.get("city", {}).get("name", "Unknown")

        for item in data.get("list", []):
            dt = datetime.fromtimestamp(item.get("dt", 0), tz=timezone.utc)
            date_key = dt.date()

            if date_key not in daily_data:
                daily_data[date_key] = {
                    "temps": [],
                    "humidity": [],
                    "conditions": [],
                    "descriptions": [],
                    "rain_prob": [],
                }

            main = item.get("main", {})
            weather = item.get("weather", [{}])[0]
            pop = item.get("pop", 0) * 100  # Probability of precipitation (0-1 -> 0-100)

            daily_data[date_key]["temps"].append(main.get("temp"))
            daily_data[date_key]["humidity"].append(main.get("humidity"))
            daily_data[date_key]["conditions"].append(weather.get("id", 800))
            daily_data[date_key]["descriptions"].append(weather.get("description"))
            daily_data[date_key]["rain_prob"].append(pop)

        # Build daily forecasts
        forecasts = []
        for i, (date, d) in enumerate(sorted(daily_data.items())[:days]):
            # Most common condition
            condition_code = max(set(d["conditions"]), key=d["conditions"].count)
            condition = self.CONDITION_MAP.get(condition_code, WeatherCondition.UNKNOWN)

            # Most common description
            description = max(set(d["descriptions"]), key=d["descriptions"].count)

            forecasts.append(
                WeatherData(
                    location=location_name,
                    temp_high=max(d["temps"]) if d["temps"] else None,
                    temp_low=min(d["temps"]) if d["temps"] else None,
                    humidity=int(sum(d["humidity"]) / len(d["humidity"]))
                    if d["humidity"]
                    else None,
                    condition=condition,
                    description=description,
                    precipitation_chance=int(max(d["rain_prob"])) if d["rain_prob"] else None,
                    forecast_date=date,
                    provider_name=self.name,
                )
            )

        return forecasts

    def _deg_to_direction(self, deg: int) -> str:
        """Convert wind degrees to cardinal direction."""
        directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
        index = round(deg / 45) % 8
        return directions[index]

    async def get_air_quality(self, latitude: float, longitude: float) -> dict:
        """
        Fetch air quality data from OpenWeatherMap Air Pollution API.

        Returns dict with: aqi, pm25, pm10, o3, no2, co, so2
        """
        url = f"https://api.openweathermap.org/data/2.5/air_pollution?lat={latitude}&lon={longitude}&appid={self.api_key}"

        try:
            with urllib.request.urlopen(url, timeout=10) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as e:
            if e.code == 401:
                raise ProviderError("OpenWeatherMap API key is invalid")
            else:
                raise ProviderError(f"Air quality API error: HTTP {e.code}")
        except Exception as e:
            raise ProviderError(f"Air quality request failed: {e}")

        return self._parse_air_quality(data)

    def _parse_air_quality(self, data: dict) -> dict:
        """Parse OpenWeatherMap air pollution response."""
        if not data.get("list"):
            return {}

        item = data["list"][0]
        components = item.get("components", {})
        main = item.get("main", {})

        # AQI is US EPA scale (1-5 mapped to 1-500 range)
        aqi_index = main.get("aqi", 1)  # 1-5 scale
        # Convert to approximate US EPA AQI scale
        aqi_map = {1: 25, 2: 75, 3: 125, 4: 175, 5: 300}
        aqi = aqi_map.get(aqi_index, 50)

        return {
            "aqi": aqi,
            "pm25": components.get("pm2_5"),
            "pm10": components.get("pm10"),
            "o3": components.get("o3"),
            "no2": components.get("no2"),
            "co": components.get("co"),
            "so2": components.get("so2"),
        }

File: src/providers/openweathermap.ts
/* eslint-disable @typescript-eslint/naming-convention */
/**
 * OpenWeatherMap provider — global fallback.
 *
 * Port of `weather/providers/openweathermap.py`. Free tier: 1000
 * calls/day. The `q={city}` query parameter accepts free-form location
 * names so no separate geocoding step is needed.
 *
 * NOTE: For parity with the Python implementation, `wind_speed` is
 * stored as m/s (the raw OWM value), even though the field comment
 * says km/h. `wind_description` is the user-facing km/h string. Do not
 * "fix" this without coordinating a Python change.
 */

import { OWM_CONDITION_MAP } from "../data-loader.js";
import { makeWeatherData } from "../models.js";
import {
  type IWeatherProvider,
  type Location,
  LocationNotSupportedError,
  ProviderError,
  WeatherCondition,
  type WeatherData,
} from "../types.js";

const BASE_URL = "https://api.openweathermap.org/data/2.5";
const AIR_POLLUTION_URL =
  "https://api.openweathermap.org/data/2.5/air_pollution";

interface AirQualityResult {
  aqi?: number;
  pm25?: number;
  pm10?: number;
  o3?: number;
  no2?: number;
  co?: number;
  so2?: number;
}

export class OpenWeatherMapProvider implements IWeatherProvider {
  readonly name = "openweathermap";
  readonly priority = 10; // Fallback priority
  readonly supportsForecast = true;
  readonly supportsAirQuality = true;
  readonly requiresApiKey = true;

  private readonly apiKey: string;

  constructor(apiKey: string) {
    this.apiKey = apiKey;
  }

  /** OpenWeatherMap supports all global locations. */
  supportsLocation(_location: Location): boolean {
    return true;
  }

  async getCurrent(location: Location): Promise<WeatherData> {
    const url = this.buildUrl("weather", {
      q: location.normalized,
      appid: this.apiKey,
      units: "metric",
    });

    const data = await this.fetchJson(url, location);

    // Optionally fetch today's forecast for high/low temps.
    let todayHigh: number | undefined;
    let todayLow: number | undefined;
    try {
      const forecastUrl = this.buildUrl("forecast", {
        q: location.normalized,
        appid: this.apiKey,
        units: "metric",
        cnt: "8", // Next 24 hours (3-hour intervals)
      });
      const forecastData = await this.fetchJson(forecastUrl, location);

      const today = new Date().toISOString().slice(0, 10); // UTC date
      const temps: number[] = [];
      for (const item of forecastData.list ?? []) {
        const dt = new Date((item.dt ?? 0) * 1000);
        const itemDate = dt.toISOString().slice(0, 10);
        if (itemDate === today) {
          const t = item.main?.temp;
          if (typeof t === "number" && Number.isFinite(t)) temps.push(t);
        }
      }
      if (temps.length > 0) {
        todayHigh = Math.max(...temps);
        todayLow = Math.min(...temps);
      }
    } catch {
      // High/low are optional enhancements.
    }

    return this.parseCurrent(data, todayHigh, todayLow);
  }

  /** Fetch current weather and merge in air-quality data. */
  async getCurrentWithAirQuality(location: Location): Promise<WeatherData> {
    const weatherData = await this.getCurrent(location);
    const { latitude: lat, longitude: lon } = weatherData;
    if (lat == null || lon == null) return weatherData;

    try {
      const aq = await this.getAirQuality(lat, lon);
      const merged: WeatherData = { ...weatherData };
      if (aq.aqi != null) merged.aqi = aq.aqi;
      if (aq.pm25 != null) merged.pm25 = aq.pm25;
      if (aq.pm10 != null) merged.pm10 = aq.pm10;
      if (aq.o3 != null) merged.o3 = aq.o3;
      if (aq.no2 != null) merged.no2 = aq.no2;
      return merged;
    } catch {
      return weatherData; // Air quality is optional.
    }
  }

  async getForecast(
    location: Location,
    days: number = 3,
  ): Promise<WeatherData[]> {
    // 5-day forecast is 3-hour intervals (40 readings max).
    const cnt = Math.min(days * 8, 40);
    const url = this.buildUrl("forecast", {
      q: location.normalized,
      appid: this.apiKey,
      units: "metric",
      cnt: String(cnt),
    });

    const data = await this.fetchJson(url, location);
    return this.parseForecast(data, days);
  }

  // ── HTTP helpers ─────────────────────────────────────────────────

  private buildUrl(path: string, params: Record<string, string>): string {
    const qs = new URLSearchParams(params).toString();
    return `${BASE_URL}/${path}?${qs}`;
  }

  private async fetchJson(
    url: string,
    location: Location,
  ): Promise<Record<string, any>> {
    let res: Response;
    try {
      res = await fetch(url);
    } catch (e) {
      throw new ProviderError(
        `OpenWeatherMap request failed: ${(e as Error).message}`,
      );
    }
    if (!res.ok) {
      if (res.status === 401) {
        throw new ProviderError("OpenWeatherMap API key is invalid");
      }
      if (res.status === 404) {
        throw new ProviderError(`Location not found: ${location.raw}`);
      }
      throw new ProviderError(
        `OpenWeatherMap API error: HTTP ${res.status}`,
      );
    }
    return (await res.json()) as Record<string, any>;
  }

  // ── Parsers ──────────────────────────────────────────────────────

  private parseCurrent(
    data: Record<string, any>,
    todayHigh?: number,
    todayLow?: number,
  ): WeatherData {
    const weather = (data.weather ?? [{}])[0] as Record<string, any>;
    const main = (data.main ?? {}) as Record<string, any>;
    const wind = (data.wind ?? {}) as Record<string, any>;
    const clouds = (data.clouds ?? {}) as Record<string, any>;
    const coords = (data.coord ?? {}) as Record<string, any>;

    const conditionCode = weather.id ?? 800;
    const condition =
      OWM_CONDITION_MAP[String(conditionCode)] ?? WeatherCondition.Unknown;

    // Wind direction
    const windDeg = typeof wind.deg === "number" ? wind.deg : undefined;
    const windDirection =
      windDeg != null ? this.degToDirection(windDeg) : undefined;

    // Format wind description (e.g. "12 km/h NE"). Python: f"{kmh:.0f}".
    let windDescription: string | undefined;
    if (typeof wind.speed === "number" && wind.speed) {
      const kmh = wind.speed * 3.6;
      const kmhRounded = Math.round(kmh).toString();
      windDescription = windDirection
        ? `${kmhRounded} km/h ${windDirection}`
        : `${kmhRounded} km/h`;
    }

    // Rain probability ≈ cloud cover (rough estimate, matches Python).
    let rainProbability: number | undefined;
    if (typeof clouds.all === "number") {
      rainProbability = Math.min(100, clouds.all);
    }

    // Capitalize description (Python: str.capitalize()).
    const rawDesc = (weather.description ?? "") as string;
    const description = rawDesc
      ? rawDesc.charAt(0).toUpperCase() + rawDesc.slice(1).toLowerCase()
      : undefined;

    // Temperature: Python's `main.get("temp")` may return None; in TS
    // we default to 0 to match the dataclass default.
    const temperature = typeof main.temp === "number" ? main.temp : 0;

    return makeWeatherData({
      location: (data.name ?? "Unknown") as string,
      temperature,
      condition,
      condition_raw: rawDesc,
      provider_name: this.name,
      observed_at: new Date(),
      ...(typeof coords.lat === "number" ? { latitude: coords.lat } : {}),
      ...(typeof coords.lon === "number" ? { longitude: coords.lon } : {}),
      ...(typeof main.feels_like === "number"
        ? { feels_like: main.feels_like }
        : {}),
      ...(todayHigh != null ? { temp_high: todayHigh } : {}),
      ...(todayLow != null ? { temp_low: todayLow } : {}),
      ...(typeof main.humidity === "number"
        ? { humidity: main.humidity }
        : {}),
      ...(typeof main.pressure === "number"
        ? { pressure: main.pressure }
        : {}),
      // For parity with Python: wind_speed stores m/s (raw OWM value),
      // not km/h. The user-facing km/h string is in wind_description.
      ...(typeof wind.speed === "number" ? { wind_speed: wind.speed } : {}),
      ...(windDirection ? { wind_direction: windDirection } : {}),
      ...(windDescription ? { wind_description: windDescription } : {}),
      ...(description ? { description } : {}),
      ...(rainProbability != null
        ? { precipitation_chance: rainProbability }
        : {}),
    });
  }

  private parseForecast(
    data: Record<string, any>,
    days: number,
  ): WeatherData[] {
    const locationName = (data.city?.name ?? "Unknown") as string;

    interface DailyBucket {
      temps: number[];
      humidity: number[];
      conditions: number[];
      descriptions: string[];
      rainProb: number[];
      date: Date;
    }

    const dailyData = new Map<string, DailyBucket>();

    for (const item of (data.list ?? []) as Array<Record<string, any>>) {
      const dt = new Date((item.dt ?? 0) * 1000);
      const dateKey = dt.toISOString().slice(0, 10); // UTC date

      let bucket = dailyData.get(dateKey);
      if (!bucket) {
        const [y, m, d] = dateKey.split("-").map(Number);
        bucket = {
          temps: [],
          humidity: [],
          conditions: [],
          descriptions: [],
          rainProb: [],
          date: new Date(Date.UTC(y!, m! - 1, d!)),
        };
        dailyData.set(dateKey, bucket);
      }

      const main = (item.main ?? {}) as Record<string, any>;
      const weather = (item.weather ?? [{}])[0] as Record<string, any>;
      const pop = ((item.pop ?? 0) as number) * 100;

      if (typeof main.temp === "number") bucket.temps.push(main.temp);
      if (typeof main.humidity === "number") bucket.humidity.push(main.humidity);
      bucket.conditions.push((weather.id ?? 800) as number);
      if (typeof weather.description === "string") {
        bucket.descriptions.push(weather.description);
      }
      bucket.rainProb.push(pop);
    }

    // Sort by date and take the first `days` entries.
    const sorted = [...dailyData.entries()]
      .sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0))
      .slice(0, days);

    const forecasts: WeatherData[] = [];
    for (const [, d] of sorted) {
      const conditionCode = mostCommon(d.conditions) ?? 800;
      const condition =
        OWM_CONDITION_MAP[String(conditionCode)] ?? WeatherCondition.Unknown;
      const description = mostCommon(d.descriptions);

      forecasts.push(
        makeWeatherData({
          location: locationName,
          temperature: 0, // Forecast entries don't have a single "current" temp.
          condition,
          provider_name: this.name,
          forecast_date: d.date,
          ...(d.temps.length > 0
            ? { temp_high: Math.max(...d.temps), temp_low: Math.min(...d.temps) }
            : {}),
          ...(d.humidity.length > 0
            ? {
                humidity: Math.round(
                  d.humidity.reduce((a, b) => a + b, 0) / d.humidity.length,
                ),
              }
            : {}),
          ...(description ? { description } : {}),
          ...(d.rainProb.length > 0
            ? { precipitation_chance: Math.round(Math.max(...d.rainProb)) }
            : {}),
        }),
      );
    }

    return forecasts;
  }

  // ── Air quality (separate endpoint) ──────────────────────────────

  async getAirQuality(
    latitude: number,
    longitude: number,
  ): Promise<AirQualityResult> {
    const url = `${AIR_POLLUTION_URL}?lat=${latitude}&lon=${longitude}&appid=${encodeURIComponent(this.apiKey)}`;

    let res: Response;
    try {
      res = await fetch(url);
    } catch (e) {
      throw new ProviderError(
        `Air quality request failed: ${(e as Error).message}`,
      );
    }
    if (!res.ok) {
      if (res.status === 401) {
        throw new ProviderError("OpenWeatherMap API key is invalid");
      }
      throw new ProviderError(`Air quality API error: HTTP ${res.status}`);
    }
    const data = (await res.json()) as Record<string, any>;
    return this.parseAirQuality(data);
  }

  private parseAirQuality(data: Record<string, any>): AirQualityResult {
    if (!Array.isArray(data.list) || data.list.length === 0) return {};

    const item = data.list[0] as Record<string, any>;
    const components = (item.components ?? {}) as Record<string, any>;
    const main = (item.main ?? {}) as Record<string, any>;

    // OWM returns AQI on a 1-5 scale; map to approximate US EPA scale.
    const aqiIndex = (main.aqi ?? 1) as number;
    const aqiMap: Record<number, number> = {
      1: 25,
      2: 75,
      3: 125,
      4: 175,
      5: 300,
    };
    const aqi = aqiMap[aqiIndex] ?? 50;

    const result: AirQualityResult = { aqi };
    if (typeof components.pm2_5 === "number") result.pm25 = components.pm2_5;
    if (typeof components.pm10 === "number") result.pm10 = components.pm10;
    if (typeof components.o3 === "number") result.o3 = components.o3;
    if (typeof components.no2 === "number") result.no2 = components.no2;
    if (typeof components.co === "number") result.co = components.co;
    if (typeof components.so2 === "number") result.so2 = components.so2;
    return result;
  }

  // ── Misc ─────────────────────────────────────────────────────────

  /** Convert wind degrees to 8-point cardinal direction. */
  private degToDirection(deg: number): string {
    const directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"];
    const index = Math.round(deg / 45) % 8;
    return directions[index]!;
  }
}

// ── Helpers ────────────────────────────────────────────────────────

/**
 * Return the most common element in an array. Mirrors Python's
 * `max(set(items), key=items.count)`. Tie-breaking matches Python's
 * iteration-order behavior: the first element seen in `Set` order
 * (= insertion order) wins.
 */
function mostCommon<T>(items: readonly T[]): T | undefined {
  if (items.length === 0) return undefined;
  const counts = new Map<T, number>();
  for (const item of items) {
    counts.set(item, (counts.get(item) ?? 0) + 1);
  }
  let best: T | undefined;
  let bestCount = -1;
  for (const [item, count] of counts) {
    if (count > bestCount) {
      best = item;
      bestCount = count;
    }
  }
  return best;
}
